Userland/lsof.cpp

```diff
@@ -0,0 +1,152 @@
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/HashMap.h>
+#include <AK/JsonArray.h>
+#include <AK/JsonObject.h>
+#include <AK/JsonParser.h>
+#include <AK/JsonValue.h>
+#include <AK/String.h>
+#include <AK/Vector.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+#include <LibCore/ProcessStatisticsReader.h>
+#include <stdio.h>
+
+struct OpenFile {
+    int fd;
+    int pid;
+    String name;
+};
+
+Vector<OpenFile> get_open_files_by_pid(pid_t pid)
+{
+    auto file = Core::File::open(String::format("/proc/%d/fds", pid), Core::IODevice::OpenMode::ReadOnly);
+    if (file.is_error()) {
+        printf("lsof: PID %d: %s\n", pid, file.error().characters());
+        return Vector<OpenFile>();
+    }
+    auto data = file.value()->read_all();
+
+    JsonParser parser(data);
+    auto result = parser.parse();
+
+    if (!result.has_value()) {
+        ASSERT_NOT_REACHED();
+    }
+
+    Vector<OpenFile> files;
+    result.value().as_array().for_each([pid, &files](const JsonValue& object) {
+        OpenFile open_file;
+        open_file.pid = pid;
+        open_file.fd = object.as_object().get("fd").to_int();
+        open_file.name = object.as_object().get("absolute_path").to_string();
+        files.append(open_file);
+    });
+    return files;
+}
+
+void display_entry(const OpenFile& file, const Core::ProcessStatistics& statistics)
+{
+    printf("%-28s %4d %4d %-10s %4d %s\n", statistics.name.characters(), file.pid, statistics.pgid, statistics.username.characters(), file.fd, file.name.characters());
+}
+
+int main(int argc, char* argv[])
+{
+    if (pledge("stdio rpath proc", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    if (unveil("/proc", "r") < 0) {
+        perror("unveil /proc");
+        return 1;
+    }
+
+    // needed by ProcessStatisticsReader::get_all()
+    if (unveil("/etc/passwd", "r") < 0) {
+        perror("unveil /proc");
+        return 1;
+    }
+
```

Comment:
You want to call `unveil(nullptr, nullptr);` after you're done setting up veiled paths, to lock the veil.